internal/upgradecheck/header.go

```diff
@@ -0,0 +1,174 @@
+}
+
+// generateHeader aggregates data and returns a struct of that data
+// If any errors are encountered, it logs those errors and uses the default values
+func generateHeader(log logr.Logger, version string, crclient crclient.Client,
+	ctx context.Context, cfg *rest.Config, isOpenShift bool) *clientUpgradeData {
+
+	var ensureDeploymentIDErr, getManagedClustersErr, getServerVersionErr error
+	var pgoClustersTotal int
+
+	upgradeInfo := &clientUpgradeData{
+		PGOVersion:  version,
+		IsOpenShift: isOpenShift,
+	}
+
+	deploymentID, ensureDeploymentIDErr = ensureDeploymentID(crclient, ctx)
+	// ensureDeploymentID always returns some ID, so log the error and continue
+	if ensureDeploymentIDErr != nil {
+		log.Error(
+			fmt.Errorf("%w", ensureDeploymentIDErr),
+			"Error in getting deploymentID, using backup",
+		)
+	}
+	upgradeInfo.DeploymentID = deploymentID
+	pgoClustersTotal, getManagedClustersErr = getManagedClusters(crclient, ctx)
+	// getManagedClusters always returns some count, so log the error and continue
+	if getManagedClustersErr != nil {
+		log.Error(
+			fmt.Errorf("%w", getManagedClustersErr),
+			"Error in getting total number of clusters",
+		)
+	}
+	upgradeInfo.PGOClustersTotal = pgoClustersTotal
+
+	serverVersion, getServerVersionErr := getServerVersion(cfg)
+	// getServerVersion always returns some version.Info, so log the error and continue
+	if getServerVersionErr != nil {
+		log.Error(
+			fmt.Errorf("%w", getServerVersionErr),
+			"Error in getting server version",
+		)
+	}
+	upgradeInfo.KubernetesEnv = serverVersion
+	return upgradeInfo
+}
+
+// ensureDeploymentID checks if the UUID exists in memory or in a ConfigMap
+// If no UUID exists, ensureDeploymentID creates one and saves it in memory/as a ConfigMap
+func ensureDeploymentID(crclient crclient.Client, ctx context.Context) (uuid.UUID, error) {
+	var err error
+	var configMapUUID uuid.UUID
+	cm := &corev1.ConfigMap{}
+
+	// If there is no deploymentID in memory, generate one for possible use
+	if deploymentID == uuid.Nil {
+		deploymentID, err = uuid.NewUUID()
+	}
+
+	// Ensure a configmap exists with a UUID
+	// If it doesn't exist, create it with the in-memory ID
+	// If it exists, use it as the in-memory ID
+	// PGO only ever creates the ConfigMap -- if it exists, it is the source of truth for the ID
+	if err == nil {
+		if err = crclient.Get(ctx, naming.AsObjectKey(
+			naming.UpgradeCheckConfigMap(),
+		), cm); err != nil {
+			if apierrors.IsNotFound(err) {
+				cm = &corev1.ConfigMap{ObjectMeta: naming.UpgradeCheckConfigMap()}
+				cm.SetGroupVersionKind(corev1.SchemeGroupVersion.WithKind("ConfigMap"))
+				cm.Data = map[string]string{
+					"deployment_id": deploymentID.String(),
+				}
+				err = errors.WithStack(crclient.Create(ctx, cm))
+				if err != nil {
+					return deploymentID, err
+				}
+
+			} else {
+				// If we encounter any error except NotFound, just go ahead with the generated ID
+				// and skip the ConfigMap
+				return deploymentID, err
```

Comment:
My thinking on this was that `ensureDeploymentID` always returns _some_ ID -- the worst case scenario would be that it just creates an ID and saves that in memory -- so the error becomes a message for the calling function. 

I figured it made sense to leave it up to the caller to decide "I cannot trust this ID because of the error" or shrug and just use the ID. (Which is how `generateHeader` uses `ensureDeploymentID` -- log the error and use whatever ID it gives.)

(That said, I don't know who would ever use this `ensureDeploymentID` func, and it would need to be changed a bit since the configmap name is hard-coded in; so there's definitely an argument to be made that I should just log the error here and return only the ID.)

Reply:
Understood - lets hash out a bit following standup, considering the thread above regarding stopping on errors.

As it stands, I personally like the idea of logging a warning/info message and not returning an error if a valid deployment ID is returned. In other words, if the function is returning some valid UUID we can/will use, then I feel as though there shouldn't be an error (I'm also not sure how the consumer would be able to tell if the UUID should be used simply based on the error message ... e.g. if I get a UUID but also some transient Kube error, what does that mean?  I.e. can I use that to determine if its a valid UUID?).

Reply:
Following discussion, we decided to stick with a standard Go pattern of logging the error rather than returning an error that is then just logged by the calling function.